enumerable.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    new_array = []
+    my_each { |x| new_array << x if yield(x) }
+    new_array
+  end
+
+  def my_all?(pat = nil)
+    answer = true
+    if block_given?
+      my_each { |x| answer = false unless yield(x) }
+    elsif pat
+      my_each { |x| answer = false unless pattern?(x, pat) }
+    else
+      my_each { |x| answer = false unless x }
+    end
+    answer
+  end
+
+  def my_any?(pat = nil)
+    answer = false
+    if block_given?
+      my_each { |x| answer = true if yield(x) }
+    elsif pat
+      my_each { |x| answer = true if pattern?(x, pat) }
+    else
+      my_each { |x| answer = true unless x }
+    end
+    answer
+  end
```

Comment:
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] hen a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern